codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -110,49 +115,82 @@ public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in)
-                    } else {
-                        composite = alloc.compositeBuffer(Integer.MAX_VALUE);
-                        composite.addComponent(true, cumulation);
+                if (cumulation instanceof CompositeByteBuf && cumulation.refCnt() == 1) {
+                    composite = (CompositeByteBuf) cumulation;
+                    if (composite.readerIndex() >= MIN_BYTES_TO_DISCARD) {
+                        composite.discardReadComponents();
+                    }
+                    // Writer index must equal capacity if we are going to "write"
+                    // new components to the end
+                    if (composite.writerIndex() != composite.capacity()) {
+                        composite.capacity(composite.writerIndex());
                     }
-                    composite.addComponent(true, in);
-                    in = null;
-                    buffer = composite;
+                } else {
+                    composite = alloc.compositeBuffer(Integer.MAX_VALUE);
+                    addFlattenedComponents(composite, cumulation);
                 }
-                return buffer;
+                addFlattenedComponents(composite, in);
+                in = null;
+                return composite;
             } finally {
                 if (in != null) {
-                    // We must release if the ownership was not transferred as otherwise it may produce a leak if
-                    // writeBytes(...) throw for whatever release (for example because of OutOfMemoryError).
+                    // We must release if the ownership was not transferred as otherwise it may produce a leak
                     in.release();
+                    if (composite != null && composite != cumulation) {
+                        composite.release();
+                    }
                 }
             }
         }
     };
 
+    /**
+     * Adds {@code toAdd} to {@code composite}. If {@code toAdd} itself is composite, it performs
+     * a "shallow copy" of the readable components. In both cases {@code composite} assumes ownership
+     * of the added buffer.
+     */
+    private static CompositeByteBuf addFlattenedComponents(CompositeByteBuf composite, ByteBuf toAdd) {
+        if (!(toAdd instanceof CompositeByteBuf)) {
+            return composite.addComponent(true, toAdd);
+        }
+        CompositeByteBuf from = (CompositeByteBuf) toAdd;
+        final int ridx = from.readerIndex();
+        final int widx = from.writerIndex();
+        final int compCount = from.numComponents();
+        int cidx = from.toComponentIndex(ridx);
+        int newReaderIdx = ridx - from.toByteIndex(cidx);
+        while (true) {
+            ByteBuf newComponent = from.internalComponent(cidx++).retain();
+            if (newReaderIdx > 0) {
+                newComponent = newComponent.slice(newReaderIdx,
+                        newComponent.readableBytes() - newReaderIdx);
```

Comment:
nit: This one could fit into a single line as well. If you are concerned about readability, please consider aligning for aesthetics' sake:

```java
newComponent = newComponent.slice(newReaderIdx,
                                  newComponent.readableBytes() - newReaderIdx);
```